php/ext/google/protobuf/message.c

```diff
@@ -1248,42 +1249,75 @@ PHP_METHOD(Timestamp, fromDateTime) {
     return;
   }
 
-  // Get timestamp from Datetime object.
-  zval retval;
-  zval function_name;
-  int64_t timestamp;
+  int64_t timestamp_seconds;
+  {
+    zval retval;
+    zval function_name;
 
 #if PHP_MAJOR_VERSION < 7
-  INIT_ZVAL(retval);
-  INIT_ZVAL(function_name);
+    INIT_ZVAL(retval);
+    INIT_ZVAL(function_name);
 #endif
 
-  PHP_PROTO_ZVAL_STRING(&function_name, "date_timestamp_get", 1);
+    PHP_PROTO_ZVAL_STRING(&function_name, "date_timestamp_get", 1);
 
-  if (call_user_function(EG(function_table), NULL, &function_name, &retval, 1,
-          ZVAL_PTR_TO_CACHED_PTR(datetime) TSRMLS_CC) == FAILURE) {
-    zend_error(E_ERROR, "Cannot get timestamp from DateTime.");
-    return;
+    if (call_user_function(EG(function_table), NULL, &function_name, &retval, 1,
+            ZVAL_PTR_TO_CACHED_PTR(datetime) TSRMLS_CC) == FAILURE) {
+      zend_error(E_ERROR, "Cannot get timestamp from DateTime.");
+      return;
+    }
+
+    protobuf_convert_to_int64(&retval, &timestamp_seconds);
+
+    zval_dtor(&retval);
+    zval_dtor(&function_name);
   }
 
-  protobuf_convert_to_int64(&retval, &timestamp);
+  int64_t timestamp_micros;
+  {
+    zval retval;
+    zval function_name;
+    zval format_string;
 
-  zval_dtor(&retval);
-  zval_dtor(&function_name);
+#if PHP_MAJOR_VERSION < 7
+    INIT_ZVAL(retval);
+    INIT_ZVAL(function_name);
+    INIT_ZVAL(format_string);
+#endif
+
+    PHP_PROTO_ZVAL_STRING(&function_name, "date_format", 1);
+    PHP_PROTO_ZVAL_STRING(&format_string, "u", 1);
+
+    CACHED_VALUE params[2] = {
+      ZVAL_PTR_TO_CACHED_VALUE(datetime),
+      ZVAL_TO_CACHED_VALUE(format_string),
+    };
+
+    if (call_user_function(EG(function_table), NULL, &function_name, &retval,
+            ARRAY_SIZE(params), params TSRMLS_CC) == FAILURE) {
```

Comment:
Instead of using ARRAY_SZIE, why not using actual number?

Reply:
See my explanation for `ARRAY_SIZE` below. 🙂
Should I change it back?

Reply:
Thanks for the explanation. ARRAY_SIZE looks good to me.